NE-1129: Make ingress operator optional on HyperShift (#1415)
openshift/enhancements: enhancements/ingress/optional-ingress-hypershift.md

- Implement the ingress operator disabling in the HyperShift's control plane operator.

## Proposal

### User Stories

- As a HyperShift engineer, I want to be able to disable the ingress operator on the clusters where the ingress is implemented differently so that the hosted control plane can become lighter.
- As a HyperShift engineer, I want the cluster operators to tolerate the absence of the ingress operator to keep the clusters from being degraded.
- As a HyperShift hosted cluster's user, I want to see the ingress capability as a known capability to bring the visibility to all the cluster actors (admins, users, other operators).
- As a HyperShift hosted cluster's user, I want the default ingress controller to not occupy the cluster resources while the ingress traffic is routed by the cloud provider's load balancers.

### Workflow Description

The HyperShift hosted cluster's user wants to use an AWS ALB to route the traffic to its example application deployed on a HyperShift's hosted cluster.

#### Terminology
[HyperShift concepts and personas](https://hypershift-docs.netlify.app/reference/concepts-and-personas/).

#### Disable the default ingress

> - Cluster service consumer requests a hosted cluster without the ingress capability using the HostedCluster API (**note**: the HostedCluster's capability API does not exist and it's out of scope of this enhancement).

Author: I suppose it'll be the same API as for the cluster version or the install config, so no way to disable explicitly, just by using the `None` baseline set. However I cannot really speak for the HyperShift engineers here.

> - Cluster service consumer requests a hosted cluster without the ingress capability using the HostedCluster API (**note**: the HostedCluster's capability API does not exist and it's out of scope of this enhancement).

Reviewer: i'm still a little confused about the mechanics overall i think, but maybe when you address my larger question from the review it will be clearer how the pieces work together.

#### Setup AWS ALB as the new ingress

- Cluster instance admin installs [AWS Load Balancer Operator](/enhancements/ingress/aws-load-balancer-operator.md).
- Cluster instance admin sets up the default `AWSLoadBalancerController`.
- Cluster instance user creates the `Ingress` resource for the default aws load balancer controller's ingress class.
- AWS load balancer controller provisions an ALB on AWS.
- Cluster instance admin can install [External DNS Operator](/enhancements/dns/externaldns-operator.md) to help cluster instance users set up convenient DNS names.

### API Extensions

N/A - no api extensions are being introduced in this proposal

### Implementation Details/Notes/Constraints

The name of the new capability is `Ingress`. It includes the cluster ingress operator and all its APIs listed below.
The implementation of the new capability follows the instructions of [how to implement a new capability](/enhancements/installer/component-selection.md#how-to-implement-a-new-capability).

#### HyperShift

HyperShift doesn't pass all the ingress operator's payload to the cluster version operator.

> Certain manifests are filtered out, most importantly: the ingress operator's deployment.

Author: It's [filtered out of the cluster ingress operator's payload today](https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L75). The deployment is then [built at runtime](https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L318).

The link to the omitted cluster ingress operator's manifest is added.

The ingress operator is considered as part of the hosted control plane, that is a resource not managed by the user hence deployed on a different cluster.
Theoretically this may question the need for the cluster capability on HyperShift. However:
- the cluster capability is a common interface for making OpenShift cluster components optional
- the cluster capability is already implemented for some operators used by HyperShift (e.g. the console operator)
- the capability provides a clear API which can be used by any party to understand whether the ingress is enabled
- [the ingress operator's APIs](#ingress-operator-apis) are still created by the cluster version operator

#### Standalone OpenShift

Making the ingress operator optional for the standalone OpenShift is a bigger scope than for HyperShift.
The changes impact the components which are not used (or used differently) in HyperShift like the installer, the authentication and monitoring operators. To prevent the scope creep for [RFE-3395](https://issues.redhat.com/browse/RFE-3395), the cluster version operator should be able to add required capabilities.

> The required capabilities should not be exposed in the capability API (install config or `ClusterVersion` object). They should be configurable in the cluster version operator and should be [implicitly enabled](https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/status.go#L161-L165) regardless of the baseline.

Reviewer: > They should be configurable in the cluster version operator

do you mean they should NOT be configurable in the CVO?

> The required capabilities should not be exposed in the capability API (install config or `ClusterVersion` object). They should be configurable in the cluster version operator and should be [implicitly enabled](https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/status.go#L161-L165) regardless of the baseline.

Reviewer: > The required capabilities should not be exposed in the capability API (install config or `ClusterVersion` object).

this seems to be in conflict w/ the suggestion below that the Console operator look at whether the ingress capability is enabled to make certain decisions.  How will the Console operator make this determination if the capability is not exposed?  The cap will at least be exposed under clusterversion.status.capabilities.enabledCapabilities, even if it can't be set under clusterversion.spec.capabilities, right?

> The required capabilities should not be exposed in the capability API (install config or `ClusterVersion` object). They should be configurable in the cluster version operator and should be [implicitly enabled](https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/status.go#L161-L165) regardless of the baseline.

Author: > i like this approach (thanks @wking who i think came up with it) but we should be clear that this isn't really implementing a new capability in the cluster. It is putting a lot of the scaffolding/backing implementation in place to make it possible to disable a component in the future, but it's not really a true capability since users can't turn it off(or on), at least in standalone (i'm still unclear what users will be able to do, or how, in hypershift).
> 
> For standalone clusters, effectively it's a read-only capability.

Yes, we can say it's not a true capability. It's a first round of the implementation.

> The required capabilities should not be exposed in the capability API (install config or `ClusterVersion` object). They should be configurable in the cluster version operator and should be [implicitly enabled](https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/status.go#L161-L165) regardless of the baseline.

Author: > > They should be configurable in the cluster version operator
> 
> do you mean they should NOT be configurable in the CVO?

By "they" I meant the required capabilities. Trevor proposed a flag for the cluster version operator. So that flag will make the required capabilities configurable.

> The required capabilities should not be exposed in the capability API (install config or `ClusterVersion` object). They should be configurable in the cluster version operator and should be [implicitly enabled](https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/status.go#L161-L165) regardless of the baseline.

Author: > The cap will at least be exposed under clusterversion.status.capabilities.enabledCapabilities, even if it can't be set under clusterversion.spec.capabilities, right?

Yes. That's exactly what I wanted to say: the required capability is not a new field in `clusterversion.spec.capabilities` but it must be in `clusterversion.status.capabilities.enabledCapabilities`.
Let me rephrase this one to avoid the confusion.

- [Route API](https://docs.openshift.com/container-platform/4.13/rest_api/network_apis/route-route-openshift-io-v1.html). Implemented in two places: OpenShift API as [CRD](https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml) and in [openshift-api-server](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route).
- [Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html). Implemented in OpenShift API as [CRD](https://github.com/openshift/api/blob/master/config/v1/0000_10_config-operator_01_ingress.crd.yaml).

However the implementation of these APIs (or parts of them) depend on the ingress operator:
- The ingress operator creates the default ingress controller to manage the backends for the routes.
- For [the component routes](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes) the ingress operator creates the required role and role bindings to get the custom TLS certificate secret from `openshift-config` namespace.

#### Console operator

> To prevent the console operator from going degraded, [the console route check](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/healthcheck/controller.go#L122) should be skipped if the ingress capability is not enabled in `ClusterVersion`.

Reviewer 2: So this is not the only usage for which we need the ingress config.
Console-operator also needs the ingress config for:
- [CLIDownloads controlle](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/clidownloads/controller.go#L118) where we use it to assemble the URL for the CLI tools
- [Routes controller](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/route/controller.go#L121) where we use it to assemble the URL for all console related route's 
- [Service controller](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/service/controller.go#L109) 

To sum it up, ingress config is used for all the logic related to console routes. Due to that reason I dont think we skip can just skip this logic, until we have other config which could substitute the ingress one.

Other option would be to disable the console capability if ingress capability is disabled.

Reviewer 3: @jhadvig If the config is present, can we just ignore that routes are not admitted and/or available as endpoints if ingress is not enabled?

Reviewer: what is all of this going to mean in terms of the console experience on a hypershift cluster that has ingress disabled?  Are we expecting the console to create LB ingress instead?

Author: It's the same problem as for any component which depends on the Route API. The long term solution is something we will need to discuss, one of the idea is described in the [Add Route API to ingress capability alternative section](https://github.com/openshift/enhancements/blob/e8779ca74822d9b425ed48e264414c2a5a8b038a/enhancements/ingress/optional-ingress-hypershift.md#add-route-api-to-the-ingress-capability).  I'm not sure the answer is known by anybody for the moment. The request is coming from ROSA, the cluster ingress operator and HyperShift are just stepping stones. The decision is likely to be taken by ROSA engineers. This EP tries to address the requirements of the RFE-3395 which [leaves this decision to the future](https://issues.redhat.com/browse/RFE-3395?focusedId=21514823&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-21514823). Also the RFE states (in [the same thread](https://issues.redhat.com/browse/RFE-3395?focusedId=21514823&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-21514823)) that AWS Load Balancer Operator will be available, so a Kubernetes service of type Load Balancer or an Ingress can do the thing eventually.

Reviewer: > It's the same problem as for any component which depends on the Route API. The long term solution is something we will need to discuss

that's fine, but this EP needs to be very clear about what the behavior/UX will be in the short term when this is disabled.  I don't think it does so yet.

Reviewer: @jhadvig can you respond to https://github.com/openshift/enhancements/pull/1415/files#r1228265620 :

> @jhadvig If the config is present, can we just ignore that routes are not admitted and/or available as endpoints if ingress is not enabled?

this seems like one of the remaining open issues on this EP before it can be approved (ensuring we understand the full implications on Console of disabling ingress, and what mitigation actions will be taken)

Author: Sorry, my bad. We discussed this with @jhadvig and reached the agreement that the console will stay running but without the ingress served by OpneShift ingress controller. The idea was similar to the one already discussed in the EP: the consumer of HyperShift (ROSA) will have to use another means to accept the ingress traffic for the console (AWS LB Operator is the candidate).

- As a HyperShift hosted cluster's user, I want to see the ingress capability as a known capability to bring the visibility to all the cluster actors (admins, users, other operators).
- As a HyperShift hosted cluster's user, I want the default ingress controller to not occupy the cluster resources while the ingress traffic is routed by the cloud provider's load balancers.

### Workflow Description

The HyperShift hosted cluster's user wants to use an AWS ALB to route the traffic to its example application deployed on a HyperShift's hosted cluster.

#### Terminology
[HyperShift concepts and personas](https://hypershift-docs.netlify.app/reference/concepts-and-personas/).

#### Disable the default ingress

- Cluster service consumer requests a hosted cluster without the ingress capability using the HostedCluster API (**note**: the HostedCluster's capability API does not exist and it's out of scope of this enhancement).
- HyperShift creates the hosted cluster.
- HyperShift skips the deployment of the ingress operator on the hosted control plane.

> - HyperShift creates `ClusterVersion` without the ingress capability on the hosted cluster.

Reviewer: This likely implies the creation of a new Capability Set that includes everything the default/all-caps set includes today, minus Ingress.

It's either that or hypershift needs to use the "None" capset, and then explicitly list all the caps except ingress, and update that list as new caps are added to the product.  (This is slightly nicer from a user perspective since it doesn't expose a capset definition to standalone OCP users, when it can't be used there, but also much less preferable from a maintenance/risk of keeping it updated perspective)

Not a serious obstacle, but worth calling out as explicit implementation requirements so we know how we're going to do it.

Author: I agree, that phrase is vague, mostly due to the lack of the vision on the future HyperShift API. I'm fine to put a new implementation detail about the creation of a dedicated "baseline sets without ingress" if the HyperShift engineering finds it necessary. However I thought of it from the conservative perceptive:  currently the only way to disable a cluster capability is by using `None` baseline - let HyperShift adapt to it. Also, a new "capability set without ingress" may be more convenient for HyperShift but more puzzling for the standalone users. I can envision a scenario where users decide to try out a new 'baseline set without the ingress' and, to their surprise, find out that the ingress was added by CVO it the end.

@csrwng: Do you have any idea of how you would like a cluster capability to be disabled? Note that currently the cluster version's API doesn't allow to exclude capabilities from the baselines. So if you the HyperShift's capability API will look exactly like cluster verison's one - the user will have to use `None` baseline and add any needed capablities except `Ingress`.

Reviewer 3: Hi @alebedev87 it sounds fine to not introduce a new capability set for this. As you mentioned, we can handle making it optional through some other means in the HyperShift API for now.

Reviewer: @csrwng it seems like we have a reasonable approach now in which we introduce the capability, but it can't be disabled on standalone clusters(for now).  Is that ok w/ you?

  - [Gateway](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_gateways.yaml)
  - [HTTPRoute](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_httproutes.yaml)

#### Related APIs

The following APIs are left intact (see [Add Route API to the ingress capability](#add-route-api-to-the-ingress-capability) to know why) if the ingress capability is disabled:
- [Route API](https://docs.openshift.com/container-platform/4.13/rest_api/network_apis/route-route-openshift-io-v1.html). Implemented in two places: OpenShift API as [CRD](https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml) and in [openshift-api-server](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route).
- [Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html). Implemented in OpenShift API as [CRD](https://github.com/openshift/api/blob/master/config/v1/0000_10_config-operator_01_ingress.crd.yaml).

> However the implementation of these APIs (or parts of them) depend on the ingress operator:

Reviewer: might be worth stating the implication explicitly:

if someone creates a Route resource, nothing will happen.

Also what will happen if someone creates an Ingress resource that normally would have been reconciled into a Route?  I assume that functionality is also missing when the Ingress capability is disabled?  (Even if it did create the Route, the Route would do nothing, but it would be good to understand if the Route object will be created to back the Ingress object or not)

Author: Done. The links to the dedicated risks and drawbacks sections are added. I had to add the "ingress to route controller" risk section and slightly update the drawbacks for this.

Reviewer: thanks, resolving this one.

### API Extensions

N/A - no api extensions are being introduced in this proposal

### Implementation Details/Notes/Constraints

#### Ingress capability

The name of the new capability is `Ingress`. It includes the cluster ingress operator and [all its APIs](#ingress-operator-apis) but does NOT include [the related APIs](#related-apis).
The implementation of the new capability should follow the instructions of [how to implement a new capability](/enhancements/installer/component-selection.md#how-to-implement-a-new-capability), to name some:
- `Ingress` capability should be added to `openshift/api` repository.
- The ingress capability should be part of the `vCurrent` and `v4.14` baseline sets.
- The openshift api vendored dependency should be bumped into `cluster-version-operator` repository.
  - Additionally some more changes will need to be done in the cluster version operator to address [the standalone use case](#standalone-openshift).
- The openshift api vendored dependency should be bumped in `installer` repository.

>   - This will allow the users of the standalone OpenShift to use the ingress capability. While the enabling of the ingress capability is not supposed to cause any issue, the disabling may have some consequences described in [Standalone OpenShift chapter](#standalone-openshift).

Reviewer: ```suggestion
  - This will allow the users of the standalone OpenShift to see the ingress capability though they will not be able to disable it(it will be implicitly enabled even if they don't choose to enable it). While the enabling of the ingress capability is not supposed to cause any issue, the disabling may have some consequences described in [Standalone OpenShift chapter](#standalone-openshift).
```

Author: Repharsed it differently to make the change minimal. I hope I didn't loose the essence.

HyperShift doesn't use the authentication operator, the authentication server is [managed by HyperShift's control plane directly](https://github.com/openshift/hypershift/blob/7cb87788779ae928d90d0e7760c1e9359f04e58a/control-plane-operator/controllers/hostedcontrolplane/oauth/deployment.go#L123-L127).

#### Monitoring operator

No change to the monitoring operator is needed until the implementation of the ingress capability for the standalone OpenShift.
The monitoring operator skips the alert manager's route check if `ingress` cluster operator is not there.

### Risks and Mitigations

#### Partially implemented capability

The ingress capability will become a part of the capability API. Meanwhile its implementation will be incomplete due to the missing support of the standalone OpenShift. However the usage of the capability is an important stepping stone, it unifies the implementation of the optional ingress for all the OpenShift installation types.

#### Component routes

In the absence of the ingress operator, the component routes cannot be customized by the owner components due to the lack of the RBAC.

> The cluster administrator can customize a component route with a custom TLS certificate secret but the component owner won't be able to access the secret. This may be misleading for the cluster administrators. However the component routes were never fully implemented on HyperShift (authentication operator is not used).

Reviewer: what is a `component route` compared to a `route`?  (since the `route` api will be entirely non-functional on hypershift when the ingress capability is disabled, i'm confused about what this paragraph is saying)

Author: [Component Routes field from Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes):
> componentRoutes is an optional list of routes that are managed by OpenShift components that a cluster-admin is able to configure the hostname and serving certificate for. The namespace and name of each route in this list should match an existing entry in the status.componentRoutes list. To determine the set of configurable Routes, look at namespace and name of entries in the .status.componentRoutes list, where participating operators write the status of configurable routes.

They are used to customize the [web console](https://docs.openshift.com/container-platform/4.13/web_console/customizing-the-web-console.html#customizing-the-console-route_customizing-web-console) and [authentication](https://docs.openshift.com/container-platform/4.13/authentication/configuring-internal-oauth.html#customizing-the-oauth-server-url_configuring-internal-oauth) routes. 

I updated this chapter to better detail the component routes.

Reviewer: thanks....i assume the componentRoutes are implemented as Ingress resources(backed by any ingress controller on the cluster), not directly as Routes?  otherwise componentRoutes would be totally broken just by virtue of the route api being non-functional when ingress-operator is disabled.

Author: Well, yeah, I put the emphasis on the minor detail which will be broken. But so far all the component routes have been implemented as routes. So, indeed, I can shorten this chapter to say just this.

Author: The chapter [updated](https://github.com/openshift/enhancements/commit/78a5a1083bdc32f62f0c2ef23ac8c9c77f538d62) to mention the absence of the route implementation.

Reviewer: thanks, resolving this one

#### Standalone OpenShift

Making the ingress operator optional for the standalone OpenShift is a bigger scope than for HyperShift.
The changes impact the components which are not used (or used differently) in HyperShift like the installer, the authentication and the monitoring operators.
To prevent the scope creep for [RFE-3395](https://issues.redhat.com/browse/RFE-3395), the following approach is proposed:
- the cluster version operator should be able to add required capabilities.
- the required capabilities is a configuration option of the cluster version operator (e.g. `--required-capabilities` flag).
- the cluster version operator should [implicitly enable](https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/status.go#L161-L165) all the provided required capabilities so that they appear in the status of `ClusterVersion` object.
- the required capabilities should be added to [the cluster version operator manifests](https://github.com/openshift/cluster-version-operator/blob/master/install/0000_00_cluster-version-operator_03_deployment.yaml) only.
- the ingress capability should be required.

> This approach should make any required capability read only on the standalone OpenShift. Also, it doesn't let the required capabilities to be set on HyperShift because HyperShift doesn't use the cluster version operator's deployment manifest but [manages the operator's deployment directly](https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L318).

Reviewer: ```suggestion
This approach should make any required capability read only on the standalone OpenShift. Also, it doesn't let capabilities be set as required on HyperShift because HyperShift doesn't use the cluster version operator's deployment manifest but [manages the operator's deployment directly](https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L318), so the required capabilities list argument won't be supplied on HyperShift.
```

- For [the component routes](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes) the ingress operator creates the required role and role binding to get the custom TLS certificate secret from the `openshift-config` namespace. See [the component routes risk](#component-routes) for further implications.

#### Console operator

In case the ingress capability is not enabled in `ClusterVersion`, the following issues arise with the console components:
- The console operator becomes degraded due to the failed [console route check](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/healthcheck/controller.go#L122).
- The console route is not served by the default ingress controller.

To resolve the former problem, it is necessary to skip [the console route check](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/healthcheck/controller.go#L122) when the ingress capability is not enabled.

> The later issue is discussed as a general case of the missing default ingress controller in [the drawbacks](#drawbacks).

Reviewer 2: Another issue I can see from the console-operator's codebase is that the operator can get degraded if the `console` route wont get admitted. The thing is that we are using the [IngressURI](https://github.com/openshift/library-go/blob/master/pkg/route/routeapihelpers/routeapihelpers.go#L18) in some of the controllers, which is checking if the route has been admitted or not.

Author: The console operator chapter is [updated](https://github.com/openshift/enhancements/commit/a739c23e48fca39cdf126eb23e3d06fcc1b14304) to mention all the route admission checks which use `IngressURI`.

I believe that was the last point to detail. After the discussion we clarified [the Ingress Config API's point](https://github.com/openshift/enhancements/pull/1415#discussion_r1228166058): it won't break the console operator as this API stays intact.

Similar to any other capability, the ingress capability:
- can be enabled not only as part of the capability sets (`vCurrent`, `v4,14`) but explicitly by using [the additional enabled capability field](https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go#L433-L438).
- if enabled, should appear in [the list of the enabled capabilities](https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go#L445-L448) of `ClusterVersion` object.
- if enabled, cannot be disabled post install.

#### HyperShift

HyperShift doesn't use the OpenShift installer to spawn the user clusters. HyperShift deploys the cluster version operator to manage the cluster operators though.
However not all the cluster operators are of use for HyperShift.
The functionality of some cluster operators is fully replaced by the logic in the hosted control plane (e.g. authentication operator).
The ingress operator is an example of a cluster operator whose life cycle is shared between the cluster version operator and HyperShift.
HyperShift doesn't pass all the ingress operator's payload to the cluster version operator.

> Certain manifests are filtered out, most importantly: [the ingress operator's deployment](https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L75).

Reviewer 4: nit: floating this link with `/main/` leaves the `L75` reference likely to go stale.  Probably worth pinning to a specific commit like https://github.com/openshift/hypershift/blob/0a5ec529ce7ae6f238ed27be6eb5757cf70c41f9/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L75 , so folks could see you were referring to `manifestsToOmit`'s `0000_50_cluster-ingress-operator_02-deployment-ibm-cloud-managed.yaml`, and then if  necessary hunt through HyperShift's subsequent Git history to figure out if/how that functionality is handled in the future.

Author: [Done](https://github.com/openshift/enhancements/commit/af34c77bc9dc3039bc58114573ccf4c90ddbf8ac), for all the links which use `#L` references.

---
title: optional-ingress-hypershift
authors:
  - "@alebedev87"
reviewers:
  - "@Miciah"
  - "@bparees"
  - "@cewong"
  - "@wking"
approvers:
  - "@Miciah"
  - "@bparees"
  - "@wking"
api-approvers:
  - N/A
creation-date: 2023-06-02
last-updated: 2023-06-02
tracking-link:
  - https://issues.redhat.com/browse/NE-1129
see-also:
  - "/enhancements/installer/component-selection.md"
  - "/enhancements/ingress/aws-load-balancer-operator.md"
replaces:
superseded-by:
---

# Make Ingress Operator optional on HyperShift

## Summary

This enhancement describes the implementation of a new capability to enable/disable the default ingress of the OpenShift cluster.

> The new capability (from here on called the ingress capability) includes [the cluster ingress operator](https://github.com/openshift/cluster-ingress-operator) as well as the custom resources managed by the operator.

Author: Applied the suggestion with some modifications.

> Would it be fair to say that the main goal is to allow the cluster admin not to have a managed router?

Well yes. By "default ingress" I meant "the ingress implementation provided by OpenShift out of the box".

This enhancement focuses on the requirements for [RFE-3395](https://issues.redhat.com/browse/RFE-3395) which targets the ROSA clusters spawned and managed by HyperShift. The implementation of the ingress capability on the standalone OpenShift is out of the scope of this enhancement.

## Motivation

Users of the ROSA managed OpenShift service would like to use [AWS load balancers](/enhancements/ingress/aws-load-balancer-operator.md) as a way to accept the ingress traffic into the cluster.
The cluster ingress operator becomes unnecessary and can be disabled to:
- reduce the resource consumption on

>   - the HyperShift client clusters (ingress controllers)

Author: I wanted to make it more reader friendly and differentiate the user and the control plane clusters. `client` is indeed not the best wording, but `hosted` 1) looks too similar to `hosted control plane`, 2) may be confused with `HostedCluster` CR, from [the HyperShift concepts and persons doc](https://hypershift-docs.netlify.app/reference/concepts-and-personas/):
>Hosted Cluster
> An OpenShift API endpoint managed by HyperShift.

I changed it to `user clusters`.

>   - the HyperShift client clusters (ingress controllers)

Author: I think I'd like to push back on the substitution of `ingress controller` with `router`. [The OCP docs](https://docs.openshift.com/container-platform/4.13/networking/ingress-operator.html) use the term `ingress controller` for the operand managed by the ingress operator:

>   - the HyperShift client clusters (ingress controllers)

Reviewer: imho the docs are wrong and we should fix them.

there are three pieces:

1) ingress operator - responsible for handling many aspects of ingress behavior/configuration on the cluster, including deploying/managing the controller for a particular ingress implementation (the route based ingress controller implementation, as well as deploying/managing the routers that handle the traffic for that ingress implementation(?)

2) ingress controller - a specific ingress implementation that uses routes to implement the ingress api.  Creates Routes from Ingress objects, and also configures the router based on Route objects.

3) router - the component responsible for handling traffic associated w/ Route objects.

so imho this sentence in our docs is wrong:

The Ingress Operator makes it possible for external clients to access your service by deploying and managing one or more HAProxy-based [Ingress Controllers](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/) to handle routing.

if you look at the page that links to:
https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/

all the "ingress controllers" being described there are not the actual traffic handling piece, they are controller that know how to talk to a traffic handling piece to configure it.  e.g.:

[AKS Application Gateway Ingress Controller](https://docs.microsoft.com/azure/application-gateway/tutorial-ingress-controller-add-on-existing?toc=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fazure%2Faks%2Ftoc.json&bc=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fazure%2Fbread%2Ftoc.json) is an ingress controller that configures the [Azure Application Gateway](https://docs.microsoft.com/azure/application-gateway/overview).

and

The [NGINX Ingress Controller for Kubernetes](https://www.nginx.com/products/nginx-ingress-controller/) works with the [NGINX](https://www.nginx.com/resources/glossary/nginx/) webserver (as a proxy).

>   - the HyperShift client clusters (ingress controllers)

Reviewer: tldr the router is not the ingress controller.  the ingress controller is one of the ingress operator's operands, however.  (and the routers are another)

  - the HyperShift hosted control plane (operator)
- make the HyperShift hosted control plane lighter and simpler, therefore more manageable
- enable the HyperShift cluster users to make their choice for the cluster ingress

### Goals
- Implement the new ingress capability.
- Add the ingress capability to the ingress operator's payload.

> - Make the dependent cluster operators tolerate the absence of the ingress operator.

Author: The absence of the ingress operator implies the absence of all the managed APIs and the objects which are created off those APIs . So I didn't want to give to much details at this point and let the rest of the EP clarify better what will be missing when the ingress operator is not there.

> - Make the dependent cluster operators tolerate the absence of the ingress operator.

Reviewer 5: It isn't obvious to me what this goal means.  The other operators shouldn't be directly impacted in any way by the ingress operator's absence, but they would be affected if the router doesn't admit the route or doesn't accept connections for it.   I suggested adding "and router" in an attempt to make it a little clearer.    If that is too concrete, how about the following?  
```
- Ensure all other cluster operators remain healthy even when the ingress capability is turned off.  
```

> - Make the dependent cluster operators tolerate the absence of the ingress operator.

Author: Suggestion applied.

- Do not lock the subsequent implementation for the standalone OpenShift.

### Non-Goals
- Use a different way for enabling/disabling the ingress other than via [the cluster capability](/enhancements/installer/component-selection.md).
- Fully implement the ingress capability for the standalone OpenShift:
  - Make the OpenShift installer tolerate the absence of the default ingress controller.
  - Make the cluster authentication operator tolerate the absence of the default ingress controller.
  - Make the cluster monitoring operator tolerate the absence of the default ingress controller.

> - Disable the Route API or try to tolerate the absence of the default ingress controller.

Reviewer: i think in this context it means "don't report yourself as failing/degraded, and everything about the component should keep working except any aspect that directly depends on having a functioning Route or ingress controller"

> - Disable the Route API or try to tolerate the absence of the default ingress controller.

Author: > "don't report yourself as failing/degraded, and everything about the component should keep working except any aspect that directly depends on having a functioning Route or ingress controller"

Yes.

Rephrasing this non goal to make this clearer.

> - Disable the Route API or try to tolerate the absence of the default ingress controller.

Reviewer 5: I find the revised phrasing confusing too:

> Tolerate the unadmitted routes for any workload except for the dependent cluster operators

Are you saying that once this EP is implemented, if a HyperShift has the ingress capability turned off, and if there are unadmitted routes, then it is fine not to "tolerate" those routes?  Does that mean it is expected that creating a route could break the cluster when the ingress capability were off?  

Put another way, if you made this non-goal into a goal, what would need to be changed to "tolerate" unadmitted routes?

> - Disable the Route API or try to tolerate the absence of the default ingress controller.

Author: > Are you saying that once this EP is implemented, if a HyperShift has the ingress capability turned off, and if there are unadmitted routes, then it is fine not to "tolerate" those routes? 

No. It may be not "fine". What I meant is that it's out of the scope of this EP.

> Does that mean it is expected that creating a route could break the cluster when the ingress capability were off?

No, it's not expected that "creating a route" could break the cluster. The HyperShift cluster should be able to become ready (all the cluster operators should be `Ready` and not `Degraded`).
However, certain user workloads or addon operators may be at risk of not being exposed or potentially broken.

> Put another way, if you made this non-goal into a goal, what would need to be changed to "tolerate" unadmitted routes?

I think that the only real solution in that case would be to implement the OpenShift route support in ALBC.

- As a HyperShift hosted cluster's user, I want to see the ingress capability as a known capability to bring the visibility to all the cluster actors (admins, users, other operators).
- As a HyperShift hosted cluster's user, I want the default ingress controller to not occupy the cluster resources while the ingress traffic is routed by the cloud provider's load balancers.

### Workflow Description

The HyperShift hosted cluster's user wants to use an AWS ALB to route the traffic to its example application deployed on a HyperShift's hosted cluster.

#### Terminology
[HyperShift concepts and personas](https://hypershift-docs.netlify.app/reference/concepts-and-personas/).

#### Disable the default ingress

- Cluster service consumer requests a hosted cluster without the ingress capability using the HostedCluster API (**note**: the HostedCluster's capability API does not exist and it's out of scope of this enhancement).
- HyperShift creates the hosted cluster.
- HyperShift skips the deployment of the ingress operator on the hosted control plane.

> - HyperShift creates `ClusterVersion` without the ingress capability on the hosted cluster.

Reviewer 5: ```suggestion
- HyperShift creates a `ClusterVersion` CR without the ingress capability on the hosted cluster.
```

- Hosted cluster's cluster version operator does not deploy the cluster ingress operator's payload (CRDs, RBACs, etc.).

#### Setup AWS ALB as the new ingress

- Cluster instance admin installs [AWS Load Balancer Operator](/enhancements/ingress/aws-load-balancer-operator.md).

> - Cluster instance admin sets up the default `AWSLoadBalancerController`.

Reviewer 5: ```suggestion
- Cluster instance admin sets up the default `AWSLoadBalancerController` CR.
```

- Cluster instance user creates the `Ingress` resource for the default aws load balancer controller's ingress class.
- AWS load balancer controller provisions an ALB on AWS.

> - Cluster instance admin can install [External DNS Operator](/enhancements/dns/externaldns-operator.md) to help cluster instance users set up convenient DNS names.

Reviewer 5: The External DNS Operator does not support ingresses (yet): https://github.com/openshift/external-dns-operator/blob/0cd8bee31e64f3086f9a44e3f5e299a3a1111175/api/v1beta1/externaldns_types.go#L471-L478

Author: True, I mixed up things: at some point I wanted this example to use `Service` which would stick well with what ExternalDNS Operator supports now. Anyway that was just a minor hint which I added to tight all together. But that's actually the unnecessary info which slightly overloads the EP: removed this line.

#### HyperShift

HyperShift doesn't use the OpenShift installer to spawn the user clusters. HyperShift deploys the cluster version operator to manage the cluster operators though.
However not all the cluster operators are of use for HyperShift.
The functionality of some cluster operators is fully replaced by the logic in the hosted control plane (e.g. authentication operator).
The ingress operator is an example of a cluster operator whose life cycle is shared between the cluster version operator and HyperShift.
HyperShift doesn't pass all the ingress operator's payload to the cluster version operator.
Certain manifests are filtered out, most importantly: [the ingress operator's deployment](https://github.com/openshift/hypershift/blob/0a5ec529ce7ae6f238ed27be6eb5757cf70c41f9/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L75).
The reason for this is that the ingress operator is considered as a part of the hosted control plane hence deployed on the control plane cluster.
However [the ingress operator's APIs](#ingress-operator-apis) and some RBACs to manage those APIs are still kept in the cluster version operator's payload.

> The ingress capability comes handy here as HyperShift can leverage it to control the remaining (not filtered) payload.

Reviewer 5: You're missing the word "in".
```suggestion
The ingress capability comes in handy here as HyperShift can leverage it to control the remaining (not filtered) payload.
```

> The HyperShift's API which will handle the hosted cluster capabilities is out of the scope of this enhancement. However this enhancement is built with the assumption that the `ClusterVersion`'s capability API (spec and status) is sufficient for any capability implementation in HyperShift.

Reviewer 5: ```suggestion
HyperShift API to specify capabilities for a hosted cluster is out of the scope of this enhancement. However, this enhancement is built with the assumption that the `ClusterVersion`'s capability API (spec and status) is sufficient for any capability implementation in HyperShift.
```
I'm not sure what "sufficient" means here; do you mean that HyperShift will have some API for specifying capabilities, and the capabilities specified in that API will map directly to the existing capabilities spec and status in the `ClusterVersion` CRD?

Reviewer: sufficient in that hypershift can use the capability api to disable the capability in question

#### Standalone OpenShift

Making the ingress operator optional for the standalone OpenShift is a bigger scope than for HyperShift.

> The changes impact the components which are not used (or used differently) in HyperShift like the installer, the authentication and the monitoring operators.

Reviewer 5: Is it accurate to say that some (not all) components that HyperShift configures differently or doesn't configure at all are affected, and that those components are specifically the installer, auth operator, and monitoring operator?  
```suggestion
The changes impact some components which are not used (or are used differently) in HyperShift, namely the installer and the authentication and monitoring operators.
```

To prevent the scope creep for [RFE-3395](https://issues.redhat.com/browse/RFE-3395), the following approach is proposed:

> - the cluster version operator should be able to add required capabilities.

Reviewer 5: You can use italics to hint that you are introducing a new term that the reader isn't expected to know already.  
```suggestion
- The cluster version operator should be able to add _required capabilities_.
```
The phrase "required capabilities" makes me think that CVO will complain if the capability is not enabled.  However, if I correctly understand your description of the concept, the idea is that these "required capabilities" are enabled by CVO whether or not they are specified in the install-config.  Maybe "force-enabled capabilities" and `--force-enable=ingress` as the flag would be clearer?
```suggestion
- The cluster version operator should be able to add _force-enabled capabilities_.
```

Reviewer: i like the terminology clarification

Reviewer 4: `--force-enabled` feels aggressive, like "do this even if customer doesn't want this".  But the reason we're setting this is because the standalone customer does want this (because they want a functional cluster) even if they don't ask for `Ingress` explicitly.  Maybe we could use `--additional-implicitly-enabled-capabilities` or something?  It doesn't have to be short, because it's just going into a manifest, and not something anyone ever needs to type on the command line.

Author: I don't have a strong opinion on the naming of the flag. I'm still fine with `--required-capabilities`. Who has the final word in CVO repository?

Reviewer: OTA team, of which @wking is a member, so i'm happy to let him pick whatever he's happy with.  This isn't a user-facing flag.

Author: Btw, the naming "required" is already used in the code of the cluster version operator for [the required feature sets](https://github.com/openshift/cluster-version-operator/blob/8913d38/pkg/cvo/cvo.go#L158-L160).

Author: `--always-enable-capabilities` is the name we agreed on. Closing the conversation.

- the required capabilities is a configuration option of the cluster version operator (e.g. `--required-capabilities` flag).
- the cluster version operator should [implicitly enable](https://github.com/openshift/cluster-version-operator/blob/2fd54d18de83b617d9d30c28cfee6383e430a102/pkg/cvo/status.go#L161-L165) all the provided required capabilities so that they appear in the status of `ClusterVersion` object.
- the required capabilities should be added to [the cluster version operator manifests](https://github.com/openshift/cluster-version-operator/blob/master/install/0000_00_cluster-version-operator_03_deployment.yaml) only.
- the ingress capability should be required.

> This approach should make any required capability read only on the standalone OpenShift.
> Also, it doesn't let capabilities be set as required on HyperShift because HyperShift doesn't use the cluster version operator's deployment manifest but [manages the operator's deployment directly](https://github.com/openshift/hypershift/blob/15fe51bda33b7acede2b4b7a7c66acc2f369a8a2/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L318),

Reviewer 5: * Hyphenate the compound adjective "read-only" to disambiguate it.  
* Use "cause" instead of "let" (enabling the required capability is still permitted and can be caused using a different API on HyperShift).  
```suggestion
This approach should make any required capability read-only on standalone OpenShift.
Also, it doesn't cause the required capabilities to be set on HyperShift because HyperShift doesn't use the cluster version operator's deployment manifest but [manages the operator's deployment directly](https://github.com/openshift/hypershift/blob/15fe51bda33b7acede2b4b7a7c66acc2f369a8a2/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L318),
```

so the required capabilities list argument won't be supplied on HyperShift.

#### Ingress operator APIs

The following APIs are not available if the ingress capability is disabled:

> - CRDs which are part of the cluster operator's manifest payload:

Reviewer 5: ```suggestion
- CRDs which are part of the cluster ingress operator's manifest payload:
```

  - [IngressController](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition.yaml)
  - [DNSRecord](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition-internal.yaml)

> - CRDs which are created at runtime if `gateway-api` feature gate is set:
>   - [GatewayClass](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_gatewayclasses.yaml)
>   - [Gateway](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_gateways.yaml)
>   - [HTTPRoute](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_httproutes.yaml)

Reviewer 5: * Might as well use the active voice to make it explicit that the ingress operator is the thing that normally creates these CRDs.
* The feature gate is named "GatewayAPI", not "gateway-api".  
* Might as well mention referencegrants for completeness.  
* Use permalinks when possible.  
```suggestion
- CRDs which the cluster ingress operator creates at runtime if the `GatewayAPI` feature gate is enabled:
  - [GatewayClass](https://github.com/openshift/cluster-ingress-operator/blob/d319e461f1c74a80474804a447237a8a679c4abd/assets/gateway-api/gateway.networking.k8s.io_gatewayclasses.yaml)
  - [Gateway](https://github.com/openshift/cluster-ingress-operator/blob/d319e461f1c74a80474804a447237a8a679c4abd/assets/gateway-api/gateway.networking.k8s.io_gateways.yaml)
  - [HTTPRoute](https://github.com/openshift/cluster-ingress-operator/blob/d319e461f1c74a80474804a447237a8a679c4abd/assets/gateway-api/gateway.networking.k8s.io_httproutes.yaml)
  - [ReferenceGrant](https://github.com/openshift/cluster-ingress-operator/blob/d319e461f1c74a80474804a447237a8a679c4abd/assets/gateway-api/gateway.networking.k8s.io_referencegrants.yaml)
```

#### Related APIs

The following APIs are left intact (see [Add Route API to the ingress capability](#add-route-api-to-the-ingress-capability) to know why) if the ingress capability is disabled:

> - [Route API](https://docs.openshift.com/container-platform/4.13/rest_api/network_apis/route-route-openshift-io-v1.html). Implemented in two places: OpenShift API as [CRD](https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml) and in [openshift-api-server](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route).

Reviewer 5: I don't see how the CRD is relevant for this EP.  
```suggestion
- [Route API](https://docs.openshift.com/container-platform/4.13/rest_api/network_apis/route-route-openshift-io-v1.html), which is [implemented in openshift-apiserver](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route).
```

Author: I think it's still relevant. If the Route API would be a part of the ingress capability the Route CRD would have to be annotated with it.

- [Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html). Implemented in OpenShift API as [CRD](https://github.com/openshift/api/blob/master/config/v1/0000_10_config-operator_01_ingress.crd.yaml).
- [Ingress API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#ingress-v1-networking-k8s-io). Implemented in Kubernetes.

However the implementation of these APIs (or parts of them) depend on the ingress operator:

> - The ingress operator creates the default ingress controller to manage the backends for the routes. See [the drawbacks](#drawbacks) for further implications.

Reviewer 5: The phrase "backends for routes" doesn't seem right.  
```suggestion
- The ingress operator creates the default ingress controller and router deployment to update routes' statuses and to proxy traffic for routes. See [the drawbacks](#drawbacks) for further implications.
```

- For [the component routes](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes) the ingress operator creates the required role and role binding to get the custom TLS certificate secret from the `openshift-config` namespace. See [the component routes risk](#component-routes) for further implications.

#### Console operator

In case the ingress capability is not enabled in `ClusterVersion`, the following issues arise with the console components:
- The console operator becomes degraded due to the failed admission checks of the console and download routes, see [the usage](https://github.com/search?q=repo%3Aopenshift%2Fconsole-operator%20IngressURI&type=code) of [IngressURI function](https://pkg.go.dev/github.com/openshift/library-go/pkg/route/routeapihelpers#IngressURI).
- The console route is not served by the default ingress controller.

To resolve the former problem, it is necessary to skip the route admission checks when the ingress capability is not enabled.   

> The later issue is discussed as a general case of the missing default ingress controller in [the drawbacks](#drawbacks).

Reviewer 5: Typo correction:
```suggestion
The latter issue is discussed as a general case of the missing default ingress controller in [the drawbacks](#drawbacks).
```

#### Authentication operator

No change to the authentication operator is needed until the implementation of the ingress capability for the standalone OpenShift.
HyperShift doesn't use the authentication operator, the authentication server is [managed by HyperShift's control plane directly](https://github.com/openshift/hypershift/blob/7cb87788779ae928d90d0e7760c1e9359f04e58a/control-plane-operator/controllers/hostedcontrolplane/oauth/deployment.go#L123-L127).

#### Monitoring operator

No change to the monitoring operator is needed until the implementation of the ingress capability for the standalone OpenShift.

> The monitoring operator skips the alert manager's route check if `ingress` cluster operator is not there.

Reviewer 5: ```suggestion
The monitoring operator skips the alert manager's route check if the `ingress` ClusterOperator CR is not present on the cluster.
```
Hm, why isn't the "ingress" clusteroperator present on a HyperShift hosted cluster?  It's in [the ingress operator's manifests](https://github.com/openshift/cluster-ingress-operator/blob/d319e461f1c74a80474804a447237a8a679c4abd/manifests/03-cluster-operator.yaml); [the CVO documentation](https://github.com/openshift/enhancements/blob/d09fbc431dcab82e730641900d5c97571b992153/dev-guide/cluster-version-operator/dev/clusteroperator.md) says, "ClusterVersionOperator will pre-create the ClusterOperator resource found in the `/manifests` folder"; and I don't see it in [`manifestsToOmit`](https://github.com/openshift/hypershift/blob/af2f0b13103f9ac1373b4abcd0b83a2870dac81c/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L47-L98).

### Risks and Mitigations

#### Partially implemented capability

> The ingress capability will become a part of the capability API. Meanwhile its implementation will be incomplete due to the missing support of the standalone OpenShift. However the usage of the capability is an important stepping stone, it unifies the implementation of the optional ingress for all the OpenShift installation types.

Reviewer 5: It seems to me that the main reason to use the capabilities API for this EP is that it will result in a consistent API for HyperShift and standalone OpenShift, and the fact that this approach results in a unified implementation is a nice side-effect.  
```suggestion
With this enhancement, the ingress capability will become a part of the capability API, but its implementation will be incomplete due to the missing support for standalone OpenShift. However, the usage of the capability API is an important stepping stone for providing a unified implementation and consistent API to make ingress optional for all the OpenShift installation types.
```

The console and the authentication operators implement the component routes by watching [the Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html) and updating the corresponding routes.    
The ingress operator creates the required role and role binding for the OpenShift component's service account so that it can get the custom service certificate secret from the `openshift-config` namespace.  
In the absence of the ingress operator, the component routes:
- cannot be customized by the owner components due to the lack of the RBAC
- become not implemented due to [the absence of the default ingress controller](#drawbacks)

The cluster instance administrator will still be able use [the Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html)
but the component owner (operator) won't be able to access the secret or rely on the route implementation.     
This may be misleading for the cluster instance administrator. However the component routes were never fully implemented on HyperShift.

> For instance, the authentication operator which is supposed to implement the authentication component route is not used at all.

Reviewer 5: The API contract for `spec.componetRoutes` states that the cluster-admin may only add an entry that has a corresponding entry in `status.componentRoutes`.  If the authentication operator and console operator do not add entries to the status, then there is nothing for the cluster-admin to configure.

Author: Adding the details about the API contract.

#### Ingress to route controller

[The Kubernetes Ingress API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#ingress-v1-networking-k8s-io) on OpenShift is implemented through translation into routes.
[The ingress-to-route controller](https://github.com/openshift/route-controller-manager#ingress-to-route-controller) is the component responsible for this functionality. When the default ingress controller is not present, the routes are left without serving, resulting in the ingresses also being unserved. On HyperShift this can be mitigated by [any ingress controller implementation](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/).

> In the context of this enhancement, [AWS Load Balancer Operator](/enhancements/ingress/aws-load-balancer-operator.md) is the expected implementation.

Reviewer 5: Would it be reasonable to turn off the ingress-to-route controller when the capability is not enabled?  Perhaps as follow-up work?

Reviewer 5: I guess "Add Route API to the ingress capability" does address my question, albeit obliquely:

> [...] In the situation when the ingress is disabled the choice lies in between the following options:
> 1. Keep the Route API and let the OpenShift components create routes which are not served by any ingress controller.
> 2. Keep the Route API and let the OpenShift components use an alternative ingress which supports the Route API.
> 3. Remove the Route API and leave the OpenShift components to adapt to the situation, which implies code changes to stop creating routes and to migrate to a different ingress API (e.g. Kubernetes Ingress).

I figured that "OpenShift components" referred to operators that created routes for their operands, but I understand now you also mean to include the ingress-to-route controller.  

This list sets up a false dichotomy; one unstated option would be to continue to let operators create their own routes but still turn off the ingress-to-route controller.  

> The first option seems to be the easiest (for the implementation), [...]

I suppose this (ease of implementation) is ultimately the reason why we don't turn off the ingress-to-route controller when the ingress capability is not enabled, right?  

> since the first option doesn't preclude the second one, it seems to be pragmatic to pick the first option, leaving open the possibility of implementing the second option later on.

I think it's a reasonable compromise to continue having the various operators (auth, console, monitoring) creating routes.  Those routes could even be useful as references when creating equivalent ingresses, should the cluster-admin choose to do so.  

However, leaving the ingress-to-route controller enabled seems less useful and more likely to mislead the typical project-admin who creates an ingress, observes that OpenShift generates a route for that ingress, and could be thereby misled to expect that OpenShift accept traffic for that ingress.  

I suppose there is the possibility that someone could turn off the ingress capability and then install a custom router deployment or even a third-party controller that implements the Route API, in which case the ingress-to-route controller _could_ still be useful.  That seems unlikely though.  

Anyway, I still think it would make sense to add a short statement here to explain why we keep translating ingresses to routes even on a cluster that doesn't serve routes, with a link to the "Add Route API to the ingress capability" section for the lengthier explanation.

Author: > I figured that "OpenShift components" referred to operators that created routes for their operands, but I understand now you also mean to include the ingress-to-route controller.

Yes. I believe I meant the ingres-to-route controller too.

> I suppose this (ease of implementation) is ultimately the reason why we don't turn off the ingress-to-route controller when the ingress capability is not enabled, right?

Right.

> However, leaving the ingress-to-route controller enabled seems less useful and more likely to mislead the typical project-admin who creates an ingress, observes that OpenShift generates a route for that ingress, and could be thereby misled to expect that OpenShift accept traffic for that ingress.

Yes, the translation of the ingresses to routes seems unnecessary in case the ingress capability is disabled.

Author: In order to highlight the discussed I added this phrase to "Ingress to route controller" risk chapter:

> It's also important to note that the routes generated from the ingress objects may become unnecessary if the ingress capability is disabled. However, this issue is not critical and can be considered for implementation as a follow-up item.

### Drawbacks

> In the absence of the default ingress controller, the Route API is left unserved. OpenShift components or workloads can create routes, but there is no available controller to support them. An important consequence of this is the impact on the ingress API which relies on [the ingress to route translation](#ingress-to-route-controller). All this may be misleading for the users.

Reviewer 5: One mitigation to this drawback is that the route status will reflect that no router has admitted the route.  This isn't perfect (the route owner may not think to check the route status, or may not realize that an empty status means that nothing is handling the route), but at least it's something savvy users or support staff can check for.

Author: Adding a remark about this mitigation.

N/A - no api extensions are being introduced in this proposal

#### Failure Modes
N/A - no api extensions are being introduced in this proposal

#### Support Procedures
N/A - no api extensions are being introduced in this proposal

## Implementation History
N/A

## Alternatives

### Implement the ingress capability for the standalone OpenShift

The goal of this enhancement is to satisfy the HyperShift's requirements. The implementation of the ingress capability for the standalone OpenShift brings a lot of additional complexity, unnecessary for HyperShift. Most probably the following components would need to be updated to tolerate the absence of the ingress operator:
- The openshift installer. It [checks for the admission of the console route](https://github.com/wking/openshift-installer/blob/3d49ffbff57011bbf846bf91de11526bb3543a83/cmd/openshift-install/create.go#L543) by the ingress controller.
- The authentication operator. It checks the availability of [the authentication server's](https://github.com/openshift/cluster-authentication-operator/blob/da39951a53ad95be28e32a48f278cc23f41e99a7/pkg/controllers/oauthendpoints/oauth_endpoints_controller.go#L30-L31) 

> and [console](https://github.com/openshift/cluster-authentication-operator/blob/da39951a53ad95be28e32a48f278cc23f41e99a7/pkg/controllers/configobservation/console/observe_consoleurl.go#L15) routes.

Reviewer 5: It doesn't look like that authentication operator check is looking up the route.  However, the console operator does perform health checks on its route: https://github.com/openshift/console-operator/blob/9fefc7fe89542e94ec4e5bb98b275716df51c311/pkg/console/controllers/healthcheck/controller.go

Reviewer 5: By the way, the authentication operator also expects the ingress operator to create the "router-certs" secret and the "default-ingress-cert" configmap with the default ingress wildcard certificate (see <https://github.com/openshift/enhancements/blob/d09fbc431dcab82e730641900d5c97571b992153/enhancements/ingress/default-ingress-cert-configmap.md>), which is to say there is a bit more work to do in the authentication operator besides turning off the route check.

Author: > By the way, the authentication operator also expects the ingress operator to create the "router-certs" secret and the "default-ingress-cert" configmap with the default ingress wildcard certificate, which is to say there is a bit more work to do in the authentication operator besides turning off the route check.

Yes, there are quite some dependencies between the authentication and ingress operators. However they are not relevant for this EP, as the authentication server is not managed by the authentication operator on HyperShift. So, the authentication operator is not used on HyperShift.

Reviewer 5: > Yes, there are quite some dependencies between the authentication and ingress operators. However they are not relevant for this EP, as the authentication server is not managed by the authentication operator on HyperShift. So, the authentication operator is not used on HyperShift.

This section is specifically about what would be required for standalone OpenShift though, so these details do seem relevant here.  I'm fine with some hand-waving though (that is, to do W, we would need to do X, Y, and Z, and (wave hands) maybe some more things, but X, Y, and Z are reason enough not to do W).

Author: The links to the usage of the ingress certificates added.

- it's a better fit for different OpenShift form factors: standalone, SNO, etc.
- the required capability logic pairs well with [the existing implicit enabling](https://github.com/openshift/cluster-version-operator/blob/2fd54d18de83b617d9d30c28cfee6383e430a102/pkg/cvo/status.go#L161-L165), so both are better to belong to the same repository.

### Alternative to the cluster capability to disable the ingress operator

Due to the lack of understanding of how HyperShift uses the cluster version operator to manage the ingress operator, we thought that the capability would serve only the informative purpose for the other OpenShift components like the console operator.
That brought the idea of giving up with the capability in favor of some other cluster resource (existing or not) which could fulfill the need of informing other OpenShift components.
Potentially that could save us from the difficult design and implementation decisions which the cluster capability implies. [The hypershift chapter](#hypershift) explains in details why the capability should be preferred over any other API.

### Add Route API to the ingress capability

> The Route API is a fundamental part of OpenShift, a lot of OpenShift components and user workloads rely on the routes. The impact of the Router API removal is difficult to anticipate and plan.
> Unlike the absence of the ingress operator which can be simulated without a dedicated capability, the removal of the route api needs the code changes in the [openshift-api-server](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route).

Reviewer 5: Minor copyedits.
```suggestion
The Route API is a fundamental part of OpenShift, and a lot of OpenShift components and user workloads rely on routes. The impact of removing the Route API is difficult to anticipate and plan for.
Unlike the absence of the ingress operator, which can be simulated without a dedicated capability, the removal of the Route API needs code changes in the [openshift-apiserver](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route).
```

This puts the implementation of the RFE at risk. In the situation when the ingress is disabled the choice lies in between the following options:
1. Keep the Route API and let the OpenShift components create routes which are not served by any ingress controller.

> 2. Keep the Route API and let the OpenShift components use an alternative ingress which supports the Route API (e.g. [AWS Load Balancer Controller](/enhancements/ingress/aws-load-balancer-operator.md)).

Reviewer 5: This is confusing—it seems to be saying that ALBO supports routes.

> 3. Remove the Route API and leave the OpenShift components adapt to the situation which implies the code changes to stop creating the routes and to migrate to a different ingress API (e.g. Kubernetes Ingress).

Reviewer 5: Minor copyedits.
```suggestion
3. Remove the Route API and leave the OpenShift components to adapt to the situation, which implies code changes to stop creating routes and to migrate to a different ingress API (e.g. Kubernetes Ingress).
```

The first option seems to be the easiest (for the implementation), the second is likely the most suitable for the long term needs (of the RFE this enhancement tries to address) and the third seems to be the most explicit and "honest" (with respect to the api users) but the most intrusive at the same time.
The choice is not obvious as all the options have drawbacks.    

> However the second one seems to be the better match for the ROSA users which already want to use the AWS load balancers.

Reviewer 5: As I understand it, ROSA users are specifically asking to replace routers+ELBs with ALBs.  
```suggestion
However the second one seems to be the better match for ROSA users who want to use AWS Application Load Balancers.
```

> Although the implementation of the second option requires the upstream work (AWS Load Balancer Controller doesn't support non Kubernetes APIs) and deserves a dedicated enhancement (and RFE).
> Since the first option doesn't exclude the second one it seems to be pragmatic to pick it leaving the room for the follow up.

Reviewer 5: Minor copyedits.  
```suggestion
Although the implementation of the second option would require upstream work (AWS Load Balancer Controller doesn't support non-Kubernetes APIs, such as the Route API) and deserves a dedicated enhancement (and RFE),
since the first option doesn't preclude the second one, it seems to be pragmatic to pick the first option, leaving open the possibility of implementing the second option later on.
```

### Add multiple cluster capabilities

The ingress operator supports multiple ingress options:
- Route API which is implemented by the HAProxy based ingress controller (default)
- Gateway API which is implemented using the OpenShift Service Mesh operator (feature gated)

Both of these options could have been separate capabilities, additional to the ingress capability.

> This alternative can be implemented later when the standalone OpenShift support will be needed.

Reviewer 5: This alternative can be considered once Gateway API support at least tech preview.  * grin *.

The new capability (from here on called the ingress capability) inhibits installation of [the cluster ingress operator](https://github.com/openshift/cluster-ingress-operator) and [its custom resources](#ingress-operator-apis).
This enhancement focuses on the requirements for [RFE-3395](https://issues.redhat.com/browse/RFE-3395) which targets the ROSA clusters spawned and managed by HyperShift. The implementation of the ingress capability on the standalone OpenShift is out of the scope of this enhancement.

## Motivation

Users of the ROSA managed OpenShift service would like to use [AWS load balancers](/enhancements/ingress/aws-load-balancer-operator.md) as a way to accept the ingress traffic into the cluster.
The cluster ingress operator becomes unnecessary and can be disabled to:
- reduce the resource consumption on
  - the HyperShift user clusters (ingress controllers)
  - the HyperShift hosted control plane (operator)
- make the HyperShift hosted control plane lighter and simpler, therefore more manageable
- enable the HyperShift cluster users to make their choice for the cluster ingress

### Goals

> - Implement the new ingress capability.
> - Add the ingress capability to the ingress operator's payload.
> - Make the dependent cluster operators tolerate the absence of the ingress operator.
> - Do not lock the subsequent implementation for the standalone OpenShift.

Reviewer 5: Can you clarify that the goal is to implement the functionality required _for HyperShift only_, but that we want to leave a path for implementing the functionality for standalone too?
```suggestion
- Implement the new ingress capability for use by HyperShift.
- Add the ingress capability to the ingress operator's payload.
- Make the dependent cluster operators tolerate the absence of the ingress operator.
- Avoid adding impediments to implementing the ingress capability for standalone OpenShift.
```

  - Make the OpenShift installer tolerate the absence of the default ingress controller.
  - Make the cluster authentication operator tolerate the absence of the default ingress controller.
  - Make the cluster monitoring operator tolerate the absence of the default ingress controller.
- Disable the Route API.
- Tolerate the unadmitted routes for any workload except for the dependent cluster operators (see [Goals](#goals)).
- Design the capability API for HyperShift.
- Implement the ingress operator disabling in the HyperShift's control plane operator.

## Proposal

### User Stories

- As a HyperShift engineer, I want to be able to disable the ingress operator on clusters that do not need it, so that the hosted control-plane can become lighter.
- As a HyperShift engineer, I want cluster operators to tolerate the absence of the ingress operator and default router deployment, so that clusters do not report degraded status in the absence of these components.
- As a HyperShift hosted cluster's user, I want the ingress capability to be reported as a known capability, so that cluster actors (admins, users, other operators) can check whether it is enabled for the cluster.

> - As a HyperShift hosted cluster's user, I want to avoid provisioning unnecessary infrastructure resources, such as an ingress ELB, when they are not needed.

Reviewer 5: I still don't know what a HyperShift engineer is.  I guess you mean "Cluster Service Consumer", [defined in the HyperShift concepts and personas](https://hypershift-docs.netlify.app/reference/concepts-and-personas/) as "The user empowered to request control planes, request workers, and drive upgrades or modify externalized configuration"?  I'm not sure what "HyperShift hosted cluster's user" is either.

Author: A "HyperShift engineer" carries the same meaning as "OpenShift engineer" in [the User Stories template from the guidelines](https://github.com/openshift/enhancements/blob/master/guidelines/enhancement_template.md#user-stories) but for the HyperShift product. 

"HyperShift hosted cluster's user" is indeed "Cluster Service Consumer", let me update this part.

Author: > "HyperShift hosted cluster's user" is indeed "Cluster Service Consumer", let me update this part.

Done.

### Workflow Description

A HyperShift Cluster User with a hosted cluster wants to use an AWS ALB to route the traffic to an application deployed on that cluster.

#### Terminology

> [HyperShift concepts and personas](https://hypershift-docs.netlify.app/reference/concepts-and-personas/).

Reviewer 5: This isn't a permalink.  Could you copy and paste the relevant definitions into the EP?

Author: Done.

- Hosted cluster's cluster version operator does not deploy the cluster ingress operator's payload (CRDs, RBACs, etc.).

#### Setup AWS ALB as the new ingress

- Cluster instance admin installs [AWS Load Balancer Operator](/enhancements/ingress/aws-load-balancer-operator.md).
- Cluster instance admin sets up the default `AWSLoadBalancerController` CR.
- Cluster instance user creates the `Ingress` resource for the default aws load balancer controller's ingress class.
- AWS load balancer controller provisions an ALB on AWS.

### API Extensions

N/A - no api extensions are being introduced in this proposal

### Implementation Details/Notes/Constraints

#### Ingress capability

The name of the new capability is `Ingress`. It includes the cluster ingress operator and [all its APIs](#ingress-operator-apis) but does NOT include [the related APIs](#related-apis).
The implementation of the new capability should follow the instructions of [how to implement a new capability](/enhancements/installer/component-selection.md#how-to-implement-a-new-capability), to name some:
- `Ingress` capability should be added to `openshift/api` repository.

> - The ingress capability should be part of the `vCurrent` and `v4.15` baseline sets.

Author: Done.

The reason for this is that the ingress operator is considered as a part of the hosted control plane hence deployed on the control plane cluster.
However [the ingress operator's APIs](#ingress-operator-apis) and some RBACs to manage those APIs are still kept in the cluster version operator's payload.
The ingress capability comes in handy here as HyperShift can leverage it to control the remaining (not filtered) payload.

HyperShift API to specify capabilities for a hosted cluster is out of the scope of this enhancement. However, this enhancement is built with the assumption that the `ClusterVersion`'s capability API (spec and status) is sufficient for any capability implementation in HyperShift.

#### Standalone OpenShift

Making the ingress operator optional for the standalone OpenShift is a bigger scope than for HyperShift.
The changes impact some components which are not used (or are used differently) in HyperShift, namely the installer and the authentication and monitoring operators.
To prevent the scope creep for [RFE-3395](https://issues.redhat.com/browse/RFE-3395), the following approach is proposed:

> - the cluster version operator should be able to add the capabilities which are always enabled.

Author: Updated to: "the cluster version operator should define the concept of _always enabled_ capabilities."

  - The ingress objects are implicitly managed by the default ingress controller. See [the ingress to route controller risk](#ingress-to-route-controller) for further implications.
- For [the component routes](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes) the ingress operator creates the required role and role binding to get the custom TLS certificate secret from the `openshift-config` namespace. See [the component routes risk](#component-routes) for further implications.

#### Console operator

In case the ingress capability is not enabled in `ClusterVersion`, the following issues arise with the console components:
- The console operator becomes degraded due to the failed admission checks of the console and download routes, see [the usage](https://github.com/search?q=repo%3Aopenshift%2Fconsole-operator%20IngressURI&type=code) of [IngressURI function](https://pkg.go.dev/github.com/openshift/library-go/pkg/route/routeapihelpers#IngressURI).
- The console route is not served by the default ingress controller.

> To resolve the former problem, it is necessary to skip the route admission checks when the ingress capability is not enabled.   

Author: Right, this is something we discussed with Jakub from the console team. The idea was to find all the occurrences of `IngressURI` function (mentioned in the sentence above) and make them tolerate the absence of the ingresscontroller. I rephrased the sentence above to mention this and removed "admission" in this sentence to mean: "all checks" or "any checks".

One mitigation to this drawback is that the route status will reflect that no ingress controller has admitted the route. This should be considered an indicator of the potential risks associated with using the route.

## Design Details

### Open Questions [optional]

N/A

### Test Plan
TBD

### Graduation Criteria

#### Dev Preview -> Tech Preview
N/A

#### Tech Preview -> GA
N/A

#### Removing a deprecated feature
N/A

### Upgrade / Downgrade Strategy

Same as [the component selection strategy](/enhancements/installer/component-selection.md#upgrade--downgrade-strategy).     
Note that:
- the capabilities [cannot disabled post install](/enhancements/installer/component-selection.md#capabilities-cannot-be-uninstalled)
- the capabilities can be [implicitly enabled](/enhancements/installer/component-selection.md#updates) (e.g. cluster version with `None` baseline upgrades to the version which moves some of the core payloads to a capability)

### Version Skew Strategy
N/A

### Operational Aspects of API Extensions

> N/A - no api extensions are being introduced in this proposal

Reviewer 5: This would be good place to explain how to check whether the ingress capability is enabled on a cluster.  Maybe also re-iterate that the ingress operator runs on the hosted control-plane, so checking for its presence on the guest cluster isn't sufficient.  However, the cluster-admin could check for the "ingress" clusteroperator or the various resources related to metrics and RBAC to determine whether the ingress operator is running, right?  The expected output of `oc -n openshift-ingress-operator get all` could be helpful.

Author: I created a HyperShift hosted cluster using the clusterbot. It turned out that `ClusterVersion` is available for the user (at least for the cluster admin). I added "Failure modes" and "Support procedures" chapters to show how to check the ingress capability.

#### Failure Modes
N/A - no api extensions are being introduced in this proposal

#### Support Procedures

> N/A - no api extensions are being introduced in this proposal

Reviewer 5: In the same vein as my previous comment, we now have one more thing to check (namely, whether the "ingress" capability is enabled) when users complain, "My route isn't working!".

Author: ["Failure modes" and "Support procedures" chapters added](https://github.com/openshift/enhancements/pull/1415#discussion_r1521326551).

- The openshift installer. It [checks for the admission of the console route](https://github.com/wking/openshift-installer/blob/3d49ffbff57011bbf846bf91de11526bb3543a83/cmd/openshift-install/create.go#L543) by the ingress controller.
- The authentication operator. It checks the availability of [the authentication server's](https://github.com/openshift/cluster-authentication-operator/blob/da39951a53ad95be28e32a48f278cc23f41e99a7/pkg/controllers/oauthendpoints/oauth_endpoints_controller.go#L30-L31) 
and [console](https://github.com/openshift/cluster-authentication-operator/blob/da39951a53ad95be28e32a48f278cc23f41e99a7/pkg/controllers/configobservation/console/observe_consoleurl.go#L15) routes.
- [The component routes API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#status-componentroutes). The ingress operator is [responsible for the RBAC](https://github.com/openshift/cluster-ingress-operator/blob/8f98c618c5609cf7fcb97e5c61dc1e7a30576925/pkg/operator/controller/configurable-route/controller.go#L42-L52)
which the component operators need to be able to access the custom TLS certificates from `openshift-config` namespace.

> **Note**: it's difficult to anticipate the impact of the ingress operator removal on the installation process without the actual capability implementation.

Reviewer 5: I think the key point here is that implementing the capability for HyperShift is a useful step towards implementing the capability for standalone OpenShift as well, right?

Author: Rephrased to 

> **Note**: it's difficult to anticipate the impact of the ingress operator removal on the installation process without the actual capability implementation. However, implementing the capability for HyperShift is a valuable step towards implementing the capability for standalone OpenShift as well.

### Allow standalone OpenShift to be broken if the ingress is disabled

This approach implies the implementation of the ingress capability the way it's described in [how to implement a new capability](/enhancements/installer/component-selection.md#how-to-implement-a-new-capability).
But unlike this enhancement, it doesn't have a goal to tolerate the absence of the ingress operator.
That is, anything which can break during or after the cluster installation is allowed to break.

> This approach has a clear downside of the postponed failure which questions the correctness of the capability implementation from the user's point of view.

Reviewer 5: In other words, we should avoid giving the user a knob that breaks the cluster, right?

Author: Yes. Rephrased this one to:

> This approach has a notable downside: it provides the user with a means to break the cluster, raising concerns about the correctness of the capability implementation from the user's perspective.

While having similarly poor level of the user experience the implementation in the cluster version operator has some advantages over this approach:
- it's a better fit for different OpenShift form factors: standalone, SNO, etc.
- the logic for the capabilities to be always enabled pairs well with [the existing implicit enabling](https://github.com/openshift/cluster-version-operator/blob/2fd54d18de83b617d9d30c28cfee6383e430a102/pkg/cvo/status.go#L161-L165), so both are better to belong to the same repository.

### Alternative to the cluster capability to disable the ingress operator

Due to the lack of understanding of how HyperShift uses the cluster version operator to manage the ingress operator, we thought that the capability would serve only the informative purpose for the other OpenShift components like the console operator.
That brought the idea of giving up with the capability in favor of some other cluster resource (existing or not) which could fulfill the need of informing other OpenShift components.

> Potentially that could save us from the difficult design and implementation decisions which the cluster capability implies. [The hypershift chapter](#hypershift) explains in details why the capability should be preferred over any other API.

Reviewer 5: This section confuses me.  After reading it a few times, I think you're saying this:

* If we added a custom API, then we would need to do these two things:
  1. Define a new "turn ingress off" API.  
  2. Modify HyperShift not to install the ingress operator when that API were used.  
* However, the HyperShift team intends to add capabilities support to HyperShift, which means we only need to do these two things:
  1. Define a new ingress capability, which is easier than defining a custom API.
  2. Annotate the ingress operator's CVO manifests with the new capability, which is easier than modifying HyperShift.  

Do I understand correctly?

Author: > Do I understand correctly?

Yes.
